### PR TITLE
Fixed #27991 -- Add 'obj' kwarg to InlineModelAdmin.has_add_permission()

### PR DESCRIPTION
**Pull request by icu0755**
https://code.djangoproject.com/ticket/27991

### CODE REVIEW

File: docs/ref/contrib/admin/index.txt [hunk 1]
```diff
@@ -2220,6 +2217,21 @@ The ``InlineModelAdmin`` class adds:
     inline forms. For example, this may be based on the model instance
     (passed as the keyword argument ``obj``).
 
+.. method:: InlineModelAdmin.has_add_permission(self, request, obj=None)
+
+    The ``has_add_permission`` method is given the HttpRequest and the parent ``obj``.
```
Comment (felixxm) [inline docs/ref/contrib/admin/index.txt]: Please wrap documentation at 79 characters.

File: django/contrib/admin/options.py [hunk 2]
```diff
@@ -534,11 +534,11 @@ def get_inline_instances(self, request, obj=None):
         for inline_class in self.inlines:
             inline = inline_class(self.model, self.admin_site)
             if request:
-                if not (inline.has_add_permission(request) or
+                if not (inline.has_add_permission(request, obj) or
                         inline.has_change_permission(request, obj) or
                         inline.has_delete_permission(request, obj)):
                     continue
-                if not inline.has_add_permission(request):
+                if not inline.has_add_permission(request, obj):
```
Comment (charettes) [inline django/contrib/admin/options.py]: As mentioned in the ticket we'll want to use an approach similar to a7c256cb5491bf2a77abdff01638239db5bfd9d5 to support `inline.has_add_permission(request)` for a certain time.

File: tests/modeladmin/tests.py [hunk 3]
```diff
@@ -617,7 +617,6 @@ def test_log_actions(self):
 
 
 class ModelAdminPermissionTests(SimpleTestCase):
-
```
Comment (timgraham) [inline tests/modeladmin/tests.py]: Please don't make unrelated whitespace changes.

File: docs/ref/contrib/admin/index.txt [hunk 4]
```diff
@@ -2220,6 +2217,21 @@ The ``InlineModelAdmin`` class adds:
...
+    Should return ``True`` if adding an instance is permitted, ``False`` otherwise.
+
+.. method:: InlineModelAdmin.has_change_permission(self, request, obj=None)
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: Doc changes for has_change_permission/has_delete_permission seem unrelated to the change.

Comment (icu0755) [inline docs/ref/contrib/admin/index.txt]: As mentioned in the ticket it is not clear that obj refers to the parent instance, and not the inline instance. The documentation just points to the ModelAdmin.has_add_permission, ModelAdmin.has_change_permission, ModelAdmin.has_delete_permission though the actual use case differs.

**Comment by jdufresne** [comment]
Hi @icu0755 Are you still working on this PR? I'm interested in this feature. If you're too busy to continue or have otherwise moved on, I can continue the work if you'd like. If you'd like to continue the work, that works too.

**Comment by icu0755** [comment]
Hi @jdufresne. Please feel free to continue or start from scratch. Unfortunately I don't know how to proceed further.

**Comment by jdufresne** [comment]
Thanks @icu0755 and thanks for the initial patch. I've created a new PR at #9721.
